internal/lib/labels/labels.go

```diff
@@ -0,0 +1,84 @@
+package labels
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+const (
+	maxLabelsLength   = 32
+	maxLabelsKVLength = 63
+)
+
+const (
+	keyType   = "key"
+	valueType = "value"
+)
+
+var errNilLabels = errors.New("labels is nil")
+
+var (
+	kvInvalidChar       = regexp.MustCompile(`[^a-z0-9\-_]+`)
+	kvValidStartEndChar = regexp.MustCompile(`^[a-z0-9].*[a-z0-9]$`)
+)
+
+type Labels map[string]string
+
+func (l Labels) Validate() error {
+	if l == nil {
+		return errNilLabels
+	}
+
+	if len(l) > maxLabelsLength {
+		return fmt.Errorf("labels length is more than [%d]", maxLabelsLength)
+	}
+
+	for key, value := range l {
+		if err := l.validateKV(key, keyType); err != nil {
+			return fmt.Errorf("error validating key [%s]: %w", key, err)
+		}
+
+		if err := l.validateKV(value, valueType); err != nil {
+			return fmt.Errorf("error validating value [%s] for key [%s]: %w", value, key, err)
+		}
+	}
+
+	return nil
+}
+
+func (l Labels) validateKV(s, _type string) error {
+	if s == "" {
+		return fmt.Errorf("%s is empty", _type)
+	}
+
+	if len(s) > maxLabelsKVLength {
+		return fmt.Errorf("%s length is more than [%d]", _type, maxLabelsKVLength)
+	}
+
+	if l.kvContainsInvalidChar(s) {
```

Comment:
no need to wrapped the regex match to dedicated function I guess, in this line, it should be suffice to use `kvInvalidChar.MatchString(s)` directly instead. wdyt?

Reply:
that's true, but the main intention for wrapping is to make it easier to read. while updating the regexp object is possible, it can lead to misleading naming on the variable intention.

Reply:
discussed internally